### PR TITLE
Add createdBy and assignedTo in task creation.

### LINKED ISSUE #306: Add `createdBy` and `assignedTo` in task creation
API contract [here](https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/tasks)

While creating a task using `POST /tasks` we need to add two more fields:

`createdBy`
- This is to track who created that particular task.
- This field shouldn't be sent from the frontend, it should be taken from the cookie itself on backend (reasons: I could send `elon-musk` in `createdBy` field from frontend, which is not something which we want 😉 )
- This is going to be the `userId` of the user and not the `username` since `userId` will remain constant for a user, but `username` may change
- Can be accessed via `req.userData.id`

`assignedTo`
- This will track whom the task is assigned to
- This will be coming from frontend as a string.
- Frontend will be sending `username` here in `req.body`, we need to find `userId` corresponding to the given `username` and store that in the DB (there are functions in `utils/users.js` which will help you do this)

It should be clear by now that we will be storing `userId`s everywhere in the DB. There's a catch though, when we are recieving and sending data, it HAS to be `username` -- we can't expect developers to remember each other's `userId`s 😛 

So in short, request and response will contain `username` but the DB will always contain the corresponding `userId`

### PR DESCRIPTION
**Pull request by rashid301**
Add optional assignee field to POST and PATCH /task API request
Remove ownerId field
Change currencies (neelam and dinero)

Refactor tasks model to add conversions between userId and username for following fields
- createdBy
- assigneee
- participants

Closes #306

### CODE REVIEW

**Review by swarajpure — CHANGES_REQUESTED** [review]
Great work! A lot of refactoring as well! 🔥 

Requesting some changes, please check.

File: controllers/tasks.js
Comment (rashid301) [reply controllers/tasks.js:11]: I had thought about it. But I wanted to keep the conversion logic at once place. It seems some of conversions were already happening in the model layer, so I decided to keep the rest over there too.

File: models/tasks.js
Comment (rashid301) [reply models/tasks.js:193]: The fetchUserTasks function is handling the error part

File: test/integration/tasks.test.js
Comment (rashid301) [reply test/integration/tasks.test.js:214]: There is a test below with ownerId = 'sumit' to update task. It would fail now after adding all the coversions.

File: models/tasks.js
Comment (rashid301) [reply models/tasks.js:179]: The fetchUserTasks function is handling the error part

File: controllers/tasks.js
Comment (swarajpure) [reply controllers/tasks.js:11]: Since you have made a new function, better to with userId from here itself

File: test/integration/tasks.test.js
Comment (swarajpure) [reply test/integration/tasks.test.js:214]: Got it, we could simply use the `addUser` util function in that case. 
Also, we need not pass other fields, only `username` should be sufficient

File: models/tasks.js
Comment (rashid301) [reply models/tasks.js]: should this change be part of this PR?

File: test/integration/tasks.test.js
Comment (rashid301) [reply test/integration/tasks.test.js:214]: First I tried using addUser. It didn't work for me. Then I found this example in other test and used it. I wasn't sure about required/optional fields. So just used the exact copy.

File: models/tasks.js
Comment (swarajpure) [reply models/tasks.js]: Yes, since you're adding the field, we'll also have to search for that 😅

File: test/integration/tasks.test.js
Comment (swarajpure) [reply test/integration/tasks.test.js:214]: Just pass this `user` object as an argument to the `addUser` function
Check the function definition for more info: https://github.com/Real-Dev-Squad/website-backend/blob/develop/test/utils/addUser.js

Comment (rashid301) [reply test/integration/tasks.test.js]: Tried again. `addUser` does not work for input arguments. That function seems unclear to me. But I am hesitant to make a change there since it is being used at many places. I am referring to this line. It is overwriting the input argument into a boolean

`user = user && Object.keys(user).length === 0 && user.constructor === Object`

File: utils/tasks.js
Comment (swarajpure) [inline utils/tasks.js:53]: These function names don't make much sense 🤔 
Can you please change them to something like `getUserIdsForTask` and `getUsernamesForTask`

File: test/integration/tasks.test.js
Comment (swarajpure) [reply test/integration/tasks.test.js]: I have asked ankur on dev-backend, waiting for his reply on the same

File: models/tasks.js
Comment (swarajpure) [reply models/tasks.js]: I mean let's try to keep returning objects simpler

File: utils/tasks.js
Comment (rashid301) [reply utils/tasks.js:53]: I guess it depends on developer's background. :). It is a very common practice to translate database objects to domain objects and vice versa. This usually happens in 2 common functions (toDao and fromDao).
The idea is all field translations in one place rather than creating mulitple functions for each fields.

E.g.
Even though the current functions only care about userId and usernames, suppose you have a new field with a date type, but you want to store them in DB as a timestamp, you only need to add those translations in these functions and not worry about any other changes CRUD operation changes. 

Does this help?

File: models/tasks.js
Comment (rashid301) [reply models/tasks.js]: pushed another change. :)

File: test/integration/tasks.test.js
Comment (swarajpure) [reply test/integration/tasks.test.js]: You can go ahead with model function, no problem 👍🏻

File: middlewares/validators/tasks.js
Comment (swarajpure) [inline middlewares/validators/tasks.js:23]: Can we have these `neelam` and `dinero` imported from constants? We have already a constant file for these currencies.

File: middlewares/validators/tasks.js
Comment (swarajpure) [inline middlewares/validators/tasks.js:57]: re: currencies from constants

File: test/integration/tasks.test.js
Comment (swarajpure) [inline test/integration/tasks.test.js:42]: re: constants

File: test/integration/tasks.test.js
Comment (swarajpure) [inline test/integration/tasks.test.js:62]: re: constants

File: test/integration/tasks.test.js
Comment (swarajpure) [inline test/integration/tasks.test.js:102]: re: constants

File: test/integration/tasks.test.js
Comment (swarajpure) [inline test/integration/tasks.test.js:199]: re: constants

**Review by swarajpure — CHANGES_REQUESTED** [review]
Some minor changes related to using constants. Please check.

Thank you!

**Review by swarajpure — CHANGES_REQUESTED** [review]
Since the structure of the task object is changed, can you please make that change in `swaggerDefinition.js` file?
After that, please see if there are any changes made by running the generate-api-schema script.

I guess we can merge this one after that :)

**Review by swarajpure — APPROVED** [review]
Looks good to me
Thank you @rashid301 😁

**Review by whyDontI — APPROVED** [review]
Great work, code quality is 🔥

**Pull request merged**
